### PR TITLE
Add PaymentIDRepository

### PR DESCRIPTION
**Pull request by Abban**
This adds a new schema, repository and use case
that allows donations or memberships to request a
new unique payment ID in order to create a
payment entity.

Also added is the required infrastructure for
the Doctrine annotation mapping.

Ticket: https://phabricator.wikimedia.org/T301115

### CODE REVIEW

File: src/UseCases/GetPaymentID/GetPaymentIDUseCase.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	private PaymentIDRepository $paymentIDRepository;
+
+	public function __construct( PaymentIDRepository $paymentIDRepository ) {
+		$this->paymentIDRepository = $paymentIDRepository;
+	}
+
+	public function getNewPaymentID(): int {
+		return $this->paymentIDRepository->getNewID();
+	}
+}
```
Comment (gbirke) [inline src/UseCases/GetPaymentID/GetPaymentIDUseCase.php:20]: You can drop this class and its test. Getting a payment id is an internal service of the payment domain, whereas "use cases" are the public API of a bounded context.

**Review by gbirke — CHANGES_REQUESTED** [review]
Different than I imagined, but I'm fine with using an entity. It has the benefit of automated DB setup and being database-independent. 

However, it has the following drawbacks:

- higher levels of PHPStan (see #61 ) will probably complain about the entity (nullable property without constructor and non-nullable getter.
- A table filling up with IDs (instead of having a table with exactly one entry). Not a big deal with modern hard drives and SSD
- I wanted to use the same DBMS (MySQL) for tests that test interaction with the database. It'll be harder to set up and and run in CI, but it'll catch subtle DBMS differences. That's one of the reasons (extending travis with MySQL services, adding a docker environment for local tests) why we estimated the original with 8 points.

**Comment by gbirke** [comment]
Sorry that my review comment did not include a clear call to action / what still needs to be done.
I'd be OK with minor changes:

- removing the use case class
- rebasing onto current main and seeing if there are PHPStan errors to be fixed (you may add exemptions for the nullable types in the PHP file or install the phpstan doctrine extension to see if that helps)
- Adding a comment on why the ID class exists

If you like to do more, I would be happy about one or more of the following changes:

- the database test uses mysql (which means changing docker-compose to include mysql and fiddling with the test environments (docker-compose, travis) to use that). 
- Maybe migrate from Travis to github actions (See https://ovirium.com/blog/how-to-make-mysql-work-in-your-github-actions/ for the mysql part)
- we drop the entity and use pure SQL instead, see [my gist example](https://gist.github.com/gbirke/6c7d67b3f4469dd3a05e269f8469f56b). We can create a class that creates/drops the necessary table and its entry with the Doctrine DBAL schema and query methods and include calls to that class in the test environment setup (for prod, we'd create a separate migration class).

**Review by gbirke — APPROVED** [review]
I like the changes, thanks!

**Comment by Abban** [comment]
We could modify the `getNewID()` method to cache the id, then delete the record, then return the id. MySQLs auto increment should remember the current highest index for us.

File: Makefile [hunk 2]
```diff
@@ -21,7 +21,7 @@ phpunit:
 	docker-compose run --rm --no-deps app ./vendor/bin/phpunit
 
 phpunit-with-coverage:
-	docker-compose -f docker-compose.yml -f docker-compose.debug.yml run --rm --no-deps -e XDEBUG_MODE=coverage app_debug ./vendor/bin/phpunit $(COVERAGE_FLAGS)
```
Comment (gbirke) [inline Makefile:24]: The `test` target now also needs dependencies

File: src/Domain/Model/PaymentID.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\PaymentContext\Domain\Model;
+
+/**
+ * Doctrine expects database mappings to have entities, so even though we don't
+ * use this class in production, having it makes configuring the test environment
+ * easier
```
Comment (gbirke) [inline src/Domain/Model/PaymentID.php:10]:
```suggestion
 * This class and its Doctrine mapping is only used in the test environment to quickly create and tear down the payment_id table. The production environment uses a migration to set up the table.
  * @internal
```

**Review by gbirke — CHANGES_REQUESTED** [review]
The tests fail intermittently in CI an ond my local machine (esp. when starting with fresh containers). To me, this looks like there is a small amount of time between the DB container being "up" and the DB accepting connections. If I run `docker-compose up -d` and then `make phpunit-with-coverage`, it works consistently, but that's no solution for CI.
Looks like the `depends_on` is not doing what it's supposed to do. Can you look into it?
Alternatively, you can check if adding a `start_dependencies` service like we do in fundraising app will help.

**Comment by gbirke** [comment]
I've looked further into this and found out that `depends_on` was removed for docker-compose >= 3.0. I suggest [following the official documentation](https://docs.docker.com/compose/startup-order/) and adding [wait-for-it.sh](https://github.com/vishnubob/wait-for-it) or [wait-for](https://github.com/Eficode/wait-for) to our Dockerfile, prepending the call to `./vendor/bin/phpunit` with the wait-for script. `wait-for` is probably better suited for the alpine image.

File: docker-compose.debug.yml [hunk 4]
```diff
@@ -1,11 +1,34 @@
...
       - XDEBUG_CONFIG=remote_host=${LOCAL_IP}
       - PHP_IDE_CONFIG=serverName=donation.spenden.wikimedia.de
     volumes:
       - ./:/usr/src/app
     working_dir: /usr/src/app
+    depends_on:
+      database:
+        condition: service_healthy
+
+  database:
```
[the remark this answers is not in the document]
Comment (Abban) [reply docker-compose.debug.yml:16]: This was me trying to get Docker to wait for the database service to be set up before starting the app service. `depends_on` only waits for the service to start, not for it to finish it's internal setup, which is why it fails for you randomly. We can have a chat about it in the daily.

**Pull request merged**